src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/PairReadWithRefBases.java

```diff
@@ -0,0 +1,51 @@
+package org.broadinstitute.hellbender.engine.dataflow.transforms;
+
+import com.google.cloud.dataflow.sdk.transforms.DoFn;
+import com.google.cloud.dataflow.sdk.transforms.PTransform;
+import com.google.cloud.dataflow.sdk.transforms.ParDo;
+import com.google.cloud.dataflow.sdk.values.KV;
+import com.google.cloud.dataflow.sdk.values.PCollection;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.reference.ReferenceBases;
+
+/**
+ * PairReadWithRefBases pairs the ReferenceBases overlapping the reads with each read and produces KV with read as
+ * the key.
+ *
+ * |---------- ref bases 1 ---------|      |-------------------- ref bases 2 --------------------|
+ * |-- read a --|   |---- read b ---|         |--------------- read c --------------|
+ * results in,
+ *
+ * |-- read a --|   |---- read b ---|         |--------------- read c --------------|
+ * |ref bases 1a|   |-ref bases 1b -|         |----------- ref bases 2c ------------|
+ *
+ * KV<read a, ref bases 1a>
+ * KV<read b, ref bases 1b>
+ * KV<read c, ref bases 2c>
+ * Precondition: each refbases must contain ALL of the bases for all of the GATKReads it's paired with.
+ */
```

Comment:
A suggestion for an alternate phrasing of the doc:

PairReadWithRefBases starts with a (Refbases->covered reads) mapping, and returns each read along with the reference bases for its interval.

This is completely a matter of opinion, but one thing I would insist you change is the word "overlapping": it should be **covered** because you require the reference shard to cover the read.

```
|--A----|
|--B--|
     |---C---|
```
Here A covers B. B overlaps C, but doesn't cover it.